### PR TITLE
[feature] Introduce stacked diagrams

### LINKED ISSUE #58501: Histogram bar orientation is always Up in a first diagram setup
### What is the bug or the crash?

Histograms' bar orientation is always `Up` when first setting up the diagram (i.e., starting from `No Diagram`). Changing the orientation and clicking on `Apply` does not update the histogram orientation on the map canvas.

Only when reopening the dialog, one can change the orientation and click on `Apply` to see changes.

### Steps to reproduce the issue

1. Load a point layer.
2. Open the Layer Diagram Properties dialog.
3. Set up a histogram diagram.
4. In the `Options` tab, select an option other than `Up`.
5. Click on `Apply`.
6. Histogram bars are not oriented according to 4.

### Versions

QGIS 3.38 and QGIS master.

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [ ] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

_No response_

### LINKED ISSUE #58782: QgsLayerStylingWidget::apply() always calls the apply() method of QgsLabelingWidget twice
### What is the bug or the crash?

Any widget inheriting from `QgsMapLayerConfigWidget` is getting its `apply()` method called twice.
This is specially noticeable when the `Live Update` mode is active.

### Steps to reproduce the issue

To make this evident, go to `qgslabelingwidget.cpp`, `apply()` method and add something like:

    QgsLogger::warning( QStringLiteral( "Call to labeling's apply" ) );

Use the labeling rendering properties in the panel widget and click on the `Apply` button, or use the `Live Update` mode activated. You'll see the duplicate call.

### Versions

At least, QGIS master.

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [ ] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

I'm creating this issue for reference. It'll be fixed in an upcoming PR.

### PR DESCRIPTION
**Pull request by gacarrillor**
This PR extends the QGIS diagram module allowing it to create and display **stacked diagrams**, i.e., diagrams located side by side. We've also taken the opportunity to move the diagram settings from a modal dialog to the **Styling dock widget**.

Stacked diagrams can be configured and **displayed horizontally or vertically**. The **spacing** between the so called subdiagrams can also be configured.

Stacked diagrams allow users to create complex diagrams like **population pyramids**.

Stacked diagrams can show **mixed subdiagram types** (e.g., a pie chart alongside a histogram) and users can add as many subdiagrams as they want. Moreover, subdiagrams can use **different renderer types**: e.g., one subdiagram could use a single category renderer, whereas the other could use a linearly interpolated one.

Subdiagrams can **set their own size and their own legend settings** (e.g., one subdiagram could show its categories in the legend, whereas another one could hide its own categories).

Subdiagram **ordering** is given by item ordering in the Stacked Diagram page's list. In the future, users could re-arrange subdiagram ordering by dragging and dropping subdiagrams (left as follow-up).

Note there are some shared settings like placement **settings** (e.g., over point), **that the stacked diagram will take from its first subdiagram**. Thus, other subdiagrams won't be able to set or change those specific settings, unless they become the first subdiagram in the stacked diagram. Therefore, when editing a subdiagram that is not located in the first place, we show labels to let users know why they don't see the expected widgets (e.g., placement ones) and where they can find them (i.e., in the first subdiagram).

**GUI Stacked Diagrams**

https://github.com/user-attachments/assets/7b5a55af-45cd-4231-85b9-c6031b783a3f

**GUI Single Diagram**

Before / After ([click](https://downloads.tuxfamily.org/tuxgis/tmp/borrar/single_diagrams_before_after.png) to see it in large size)
image

image

**Other samples** 
(Note the corresponding point feature's geometry)

expected_stackedhistograms

expected_stackedpieshorizontalwithspacing

image

image

Mixed renderer type (pies using single category and histograms using linearly interpolated one):
expected_stackedpiehistogram

Nested stacked diagrams (just because we can! Not exposed through GUI, but possible in core and showcased via unit test):
expected_stackeddiagramsnested

-------------
Includes tests.

Additional fixes:
 + Fix #58501
 + Fix #58782 

----------

Funded by [Landesamt für Vermessung und Geoinformation (LVG) Vorarlberg](https://vorarlberg.at/-/landesamt-fuer-vermessung-und-geoinformation-l-1) in collaboration with the [QGIS user group Switzerland](https://qgis.ch/).

### CODE REVIEW

**Comment by nyalldawson** [comment]
Nice work @gacarrillor  !

I don't think the use of tabs is appropriate here for configuring sub diagrams. I'd rather see the UI modeled off our existing rule based labeling approach, which IMO is a better fit for this kind of configuration.

**Comment by nyalldawson** [comment]
I'd also suggest that the "stacked diagram" option should appear in the topmost combo box alongside the actual diagram types (eg alongside Histogram etc). Then there's no added GUI complexity when using the common single diagram type, and selecting "stacked diagrams" would expose the list of subdiagrams in the panel below that combo.

**Comment by andreasneumann** [comment]
> Funded by the [QGIS user group Switzerland](https://qgis.ch/).

It is a co-funding between "Landesvermessungsamt Vorarlberg" (the larger share) and the Swiss QGIS user group (the smaller share). So actually, "Landesvermessungsamt Vorarlberg"  should be listed first.

**Comment by andreasneumann** [comment]
Looking forward to use this in the future - a really nice extension to the diagram system!

**Comment by gacarrillor** [comment]
> It is a co-funding between "Landesvermessungsamt Vorarlberg" (the larger share) and the Swiss QGIS user group (the smaller share). So actually, "Landesvermessungsamt Vorarlberg" should be listed first.

@andreasneumann, would

Funded by [Landesamt für Vermessung und Geoinformation (LVG)](https://vorarlberg.at/-/landesamt-fuer-vermessung-und-geoinformation-l-1) in collaboration with the [QGIS user group Switzerland](https://qgis.ch/).

work for you or should we stick to 

Funded by [Landesvermessungsamt Vorarlberg](https://vorarlberg.at/-/landesamt-fuer-vermessung-und-geoinformation-l-1) in collaboration with the [QGIS user group Switzerland](https://qgis.ch/).

or another one that you prefer?

**Comment by gacarrillor** [comment]
Thank you all for the feedback!
I'm excited about this feature!

**Comment by andreasneumann** [comment]
> Funded by [Landesvermessungsamt Vorarlberg](https://vorarlberg.at/-/landesamt-fuer-vermessung-und-geoinformation-l-1) in collaboration with the [QGIS user group Switzerland](https://qgis.ch/).

Officially it is called "Landesamt für Vermessung und Geoinformation (LVG) Vorarlberg". I think Vorarlberg should be in the name. It should also be used in the visual changelog. Thanks a lot!

File: src/core/qgsdiagramrenderer.cpp [hunk 1]
```diff
@@ -804,6 +837,255 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
...
+
+QList<QString> QgsStackedDiagramRenderer::diagramAttributes() const
+{
+  return mSettings.categoryAttributes;
+}
+
+QList< QgsLayerTreeModelLegendNode * > QgsStackedDiagramRenderer::legendItems( QgsLayerTreeLayer *nodeLayer ) const
+{
+  QList< QgsLayerTreeModelLegendNode * > nodes;
+  for ( int i = 0; i < rendererCount(); i++ )
```
Comment (m-kuhn) [inline src/core/qgsdiagramrenderer.cpp:966]: Any reason not to use a range based for loop here?

`for( QgsDiagramRenderer *renderer : mDiagramRenderers )`?

Comment (gacarrillor) [reply src/core/qgsdiagramrenderer.cpp:966]: Not really @m-kuhn, I could adjust it.

Comment (m-kuhn) [reply src/core/qgsdiagramrenderer.cpp:966]: They are normally preferable

**Comment by m-kuhn** [comment]
@gacarrillor is there anything that keeps this in "draft"? Feature freeze is in two days and this would be a fantastic feature to ship with the next LTR, so I wonder if we have any chance to see this there.

**Comment by gacarrillor** [comment]
> @gacarrillor is there anything that keeps this in "draft"? Feature freeze is in two days and this would be a fantastic feature to ship with the next LTR, so I wonder if we have any chance to see this there.

Thanks for pointing this out. I've just removed the draft mark.
I'm currently addressing the GUI/UX suggestions made by @nyalldawson, which are really convenient here.
Looking forward to seeing this landing in QGIS 3.40.

**Comment by gacarrillor** [comment]
Thank you @nyalldawson for the exemption.
While addressing your review, we felt we had a great opportunity to finally move the diagram dialog to the Style dock widget, and we're really excited about bringing this to QGIS (we're doing our best). I hope we can push by mid next week.

Preview:
image

File: src/core/qgsdiagramrenderer.cpp (hunk 1 above)
Comment (gacarrillor) [reply src/core/qgsdiagramrenderer.cpp]: Addressed in https://github.com/qgis/QGIS/commit/f81df2876c9874f251eb01d7551f443fa37e2df3#diff-c2267f7efae558b8c9bc33e2e751364580eb0b84a7b9b709d3ac815b601bf8eaR977

Is this form also welcome?

`for ( const auto &renderer : std::as_const( mDiagramRenderers ) )`

**Comment by gacarrillor** [comment]
Demo:

https://github.com/user-attachments/assets/7b5a55af-45cd-4231-85b9-c6031b783a3f

Known issues to be addressed in upcoming commits:

 + Scale dependent visibility on stacked diagrams.
 + Remove a couple of horizontal scroll bars in dock mode.
 + Replace remaining dialogs in dock widget mode.
 + GUI/UX unexpected behaviors, like default values/default selections.

Most-likely for a follow-up:

 + Drag and drop for diagrams.
 + Copy/paste of diagrams.
 + Passing from stacked to single diagram: take first stacked diagram as the single one.

-------

@nyalldawson, there are still some details to be solved, but we've almost finished addressing your suggestions, thanks!

File: src/core/diagram/qgsstackedbardiagram.h [hunk 2]
```diff
@@ -16,7 +16,7 @@
 #ifndef QGSSTACKEDBARDIAGRAM_H
 #define QGSSTACKEDBARDIAGRAM_H
 
-#define DIAGRAM_NAME_STACKED "Stacked"
+#define DIAGRAM_NAME_STACKED_BAR "StackedBar"
```
Comment (3nids) [inline src/core/diagram/qgsstackedbardiagram.h:19]: I am not 100% sure, but we usually prefer static const members over #define
https://stackoverflow.com/questions/1637332/static-const-vs-define
Might be worth switching them?

File: src/core/diagram/qgsstackeddiagram.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+{
+  return DIAGRAM_NAME_STACKED;
+}
+
+QSizeF QgsStackedDiagram::diagramSize( const QgsAttributes &attributes, const QgsRenderContext &c, const QgsDiagramSettings &s )
+{
+  Q_UNUSED( attributes )
+  Q_UNUSED( c )
+  Q_UNUSED( s )
+  return QSizeF( 0, 0 );
```
Comment (3nids) [inline src/core/diagram/qgsstackeddiagram.cpp:70]: Do you know why we have `diagramSize` implemented both in `QgsStackedDiagram` and `QgsStackedDiagramRenderer` ?

File: src/core/qgsdiagramrenderer.cpp [hunk 4]
```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
...
+  }
+
+  QPointF newPos = pos; // Each subdiagram will have its own newPos
+
+  // Get subrenderers sorted by mode (vertical diagrams are returned backwards)
+  const QList< QgsDiagramRenderer * > stackedRenderers = renderers( true );
+
+  for ( const auto &stackedRenderer : stackedRenderers )
+  {
+    if ( stackedRenderer->rendererName() == QStringLiteral( "Stacked" ) )
```
Comment (3nids) [inline src/core/qgsdiagramrenderer.cpp:921]: stacked diagrams in stacked diagrams
inception-kukkendare

File: src/core/qgsdiagramrenderer.cpp [hunk 5]
```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
...
+  Q_UNUSED( c )
+  Q_UNUSED( s )
+  return false;
+}
+
+QSizeF QgsStackedDiagramRenderer::diagramSize( const QgsFeature &feature, const QgsRenderContext &c ) const
+{
+  Q_UNUSED( feature )
+  Q_UNUSED( c )
+  return QSizeF( 0, 0 );
```
Comment (3nids) [inline src/core/qgsdiagramrenderer.cpp:990]: Why is this does needed to go through the children and return the global size?

File: src/core/qgsdiagramrenderer.cpp [hunk 6]
```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
...
+    stackedDiagram->subDiagramPosition( newPos, c, mSettings, s );
+  }
+}
+
+bool QgsStackedDiagramRenderer::diagramSettings( const QgsFeature &feature, const QgsRenderContext &c, QgsDiagramSettings &s ) const
+{
+  Q_UNUSED( feature )
+  Q_UNUSED( c )
+  Q_UNUSED( s )
+  return false;
```
Comment (3nids) [inline src/core/qgsdiagramrenderer.cpp:983]: This doesn't sound super nice to implement the methods here. But I don't see any approach to avoid this.

File: src/core/qgsdiagramrenderer.cpp [hunk 7]
```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
...
+
+    if ( !s.enabled )
+    {
+      continue;
+    }
+
+    if ( s.scaleBasedVisibility )
+    {
+      double maxScale = s.maximumScale;
+      if ( maxScale > 0 && c.rendererScale() < maxScale )
```
Comment (3nids) [inline src/core/qgsdiagramrenderer.cpp:942]: isn't one the boundaries inclusive for scale visibility?

File: src/core/qgsdiagramrenderer.cpp [hunk 8]
```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
...
+{
+  Q_UNUSED( feature )
+  Q_UNUSED( c )
+  return QSizeF( 0, 0 );
+}
+
+QList<QgsDiagramSettings> QgsStackedDiagramRenderer::diagramSettings() const
+{
+  QList<QgsDiagramSettings> settingsList;
+  settingsList.push_back( mSettings );
```
Comment (3nids) [inline src/core/qgsdiagramrenderer.cpp:996]: why do you need a list of settings and not just the settings?

File: src/core/qgsdiagramrenderer.cpp [hunk 9]
```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
...
+
+QList< QgsDiagramRenderer * > QgsStackedDiagramRenderer::renderers( bool sortByDiagramMode ) const
+{
+  QList< QgsDiagramRenderer * > renderers;
+
+  if ( sortByDiagramMode && mSettings.stackedDiagramMode == QgsDiagramSettings::Vertical )
+  {
+    // We draw vertical diagrams backwards, so
+    // we return the subrenderers in reverse order
+    QList< QgsDiagramRenderer * >::const_reverse_iterator iter = mDiagramRenderers.rbegin();
```
Comment (3nids) [inline src/core/qgsdiagramrenderer.cpp:1024]: you can use a oneliner
`std::reverse(mDiagramRenderers.begin(), mDiagramRenderers.end());`

**Review by 3nids — COMMENTED** [review]
I looked at the core part. I will do the gui tomorrow!
Looks good so far!

File: src/core/qgsdiagramrenderer.cpp (hunk 4 above)
Comment (gacarrillor) [reply src/core/qgsdiagramrenderer.cpp:921]: When I first thought about it, and gave it a try, that specific `if` clause was the only missing piece.

Note this opens the opportunity to have grids of diagrams instead of _only_ vertical or horizontal stacks.

As mentioned in the PR description, this functionality is not yet offered via GUI, but only demonstrated via unit test.

File: src/core/qgsdiagramrenderer.cpp (hunk 6 above)
Comment (gacarrillor) [reply src/core/qgsdiagramrenderer.cpp:983]: Yeap, those are pure virtual methods, so I was required to override them returning a default value, but they're never actually called for the stacked diagram renderer. That renderer is _special_ in that sense, it delegates several of their methods to sub renderers.

File: src/core/qgsdiagramrenderer.cpp (hunk 5 above)
Comment (gacarrillor) [reply src/core/qgsdiagramrenderer.cpp:990]: `diagramSize` is protected, it'll only be called inside the public method `sizeMapUnits()`, which we do override for Stack Diagrams, delegating to subrenderers and adding up their sizes to get the stacked diagram size.

So, for the Stacked Diagram Renderer, the `diagramSize()` is never called, and that's why we do a dummy implementation returning a default value (it's a pure virtual method, so we needed to override it).

File: src/core/qgsdiagramrenderer.cpp (hunk 8 above)
Comment (gacarrillor) [reply src/core/qgsdiagramrenderer.cpp:996]: That's the way the superclass is implemented (Pure virtual method as well). So far, subclasses stick to have a list of one settings element.

The rationale is probably to have the possibility to somehow extend the renderer to have more available settings.

File: src/core/diagram/qgsstackeddiagram.cpp (hunk 3 above)
Comment (gacarrillor) [reply src/core/diagram/qgsstackeddiagram.cpp:70]: It's the diagram who calculates the diagram size for a specific feature, but depending on the rendering type, the rendering subclass needs to pass additional arguments to the diagram to do that.

File: src/core/qgsdiagramrenderer.cpp (hunk 9 above)
Comment (gacarrillor) [reply src/core/qgsdiagramrenderer.cpp:1024]: Nice tip, thanks! Looks cleaner now.
I've used it in a copy instead, to avoid getting confused on what is the (normal) first diagram after switching modes (e-g-. from vertical to horizontal).

File: src/core/diagram/qgsstackedbardiagram.h (hunk 2 above)
Comment (gacarrillor) [reply src/core/diagram/qgsstackedbardiagram.h:19]: Yes, definitely worth it! Thanks for the suggestion.
I'll address this one tomorrow.

File: src/gui/vector/qgsdiagramproperties.cpp [hunk 10]
```diff
@@ -1118,3 +1245,71 @@ void QgsDiagramProperties::createAuxiliaryField()
...
 }
+
+void QgsDiagramProperties::connectValueChanged( const QList<QWidget *> &widgets, const char *signal )
+{
+  const auto constWidgets = widgets;
+  for ( QWidget *widget : constWidgets )
+  {
+    if ( QgsSymbolButton *w = qobject_cast<QgsSymbolButton *>( widget ) )
+    {
+      connect( w, SIGNAL( changed() ), this, signal );
```
Comment (3nids) [inline src/gui/vector/qgsdiagramproperties.cpp:1256]: Can you switch theses to functor based connections?

File: src/core/qgsdiagramrenderer.cpp (hunk 5 above)
Comment (m-kuhn) [reply src/core/qgsdiagramrenderer.cpp:990]: In this case, would it make sense to make it `private` instead of `protected`?

**Review by 3nids — APPROVED** [review]
Not much more comments regarding the GUI, good job!

File: src/core/qgsdiagramrenderer.cpp (hunk 5 above)
Comment (gacarrillor) [reply src/core/qgsdiagramrenderer.cpp:990]: Yeap, it's existing API. It's in the abstract superclass, it must be overridden by subclasses.

**Comment by gacarrillor** [comment]
@DelazJ, @nirvn, I'd love your feedback on the GUI/UX part if you can find some time to check it.

File: src/gui/vector/qgsdiagramproperties.cpp (hunk 10 above)
Comment (gacarrillor) [reply src/gui/vector/qgsdiagramproperties.cpp:1256]: Yes, should I also adjust https://github.com/qgis/QGIS/blob/da1f55b7d1ca1026535ebadb39303df4447cc729/src/gui/symbology/qgsrendererpropertiesdialog.cpp#L140 and https://github.com/qgis/QGIS/blob/da1f55b7d1ca1026535ebadb39303df4447cc729/src/gui/qgstextformatwidget.cpp#L662 for consistency?

File: src/core/diagram/qgsstackedbardiagram.h (hunk 2 above)
Comment (gacarrillor) [reply src/core/diagram/qgsstackedbardiagram.h]: Done in https://github.com/qgis/QGIS/pull/58568/commits/463bf249bb04c91f6cadd5cab834fd3ee836e446

File: src/gui/vector/qgsdiagramproperties.cpp (hunk 10 above)
Comment (gacarrillor) [reply src/gui/vector/qgsdiagramproperties.cpp]: Done in https://github.com/qgis/QGIS/pull/58568/commits/01055d7a2d48f999ea193410fec7e5764d5135cc and https://github.com/qgis/QGIS/pull/58568/commits/eb201c56e61919844b21659f9c3493d3a6441332

File: src/core/diagram/qgspiediagram.h [hunk 11]
```diff
@@ -38,6 +36,8 @@ class QgsRenderContext;
 class CORE_EXPORT QgsPieDiagram: public QgsDiagram SIP_NODEFAULTCTORS
 {
   public:
+    static const QString DIAGRAM_NAME_PIE;
```
Comment (3nids) [inline src/core/diagram/qgspiediagram.h:39]:
```suggestion
    static const QString DIAGRAM_NAME_PIE SIP_SKIP;
```
can you do the same for all the others?

File: src/core/qgsdiagramrenderer.cpp (hunk 7 above)
Comment (gacarrillor) [reply src/core/qgsdiagramrenderer.cpp]: Good catch! I think you've found a good source of unexpected behaviors. I had copied this block from the current `QgsVectorLayerDiagramProvider::registerDiagram()`.

After some GUI tests with a given scale (say 1:50.000), the scale is sometimes a round number and sometimes a number like `49999.999999998996` or `50000.000000003354` (I could reproduce it when maximizing/restoring the QGIS main window), so those checks were failing from time to time, which could make the diagrams disappear when they're expected to be rendered and vice versa. 

I added checks using qgsDoubleNear() to take non-round values into account.
Applied in https://github.com/qgis/QGIS/pull/58568/commits/b8f976012ab383e29e7156ff75ea3ea836b99e06

**Comment by gacarrillor** [comment]
Thanks a lot @3nids for your insightful review!

File: src/core/diagram/qgsstackedbardiagram.cpp [hunk 12]
```diff
@@ -117,7 +119,7 @@ double QgsStackedBarDiagram::legendSize( double value, const QgsDiagramSettings
 
 QString QgsStackedBarDiagram::diagramName() const
 {
-  return DIAGRAM_NAME_STACKED;
+  return QgsStackedBarDiagram::DIAGRAM_NAME_STACKED_BAR;
```
Comment (nyalldawson) [inline src/core/diagram/qgsstackedbardiagram.cpp:122]: @gacarrillor this has broken restoring old projects using stacked bar diagrams -- you'll need to revert this change, and then introduce a new unique, different string for the stacked diagram type.

Comment (gacarrillor) [reply src/core/diagram/qgsstackedbardiagram.cpp:122]: Thanks for the report and sorry.
Solved in #58915

File: src/core/qgsdiagramrenderer.cpp [hunk 13]
```diff
@@ -804,6 +857,273 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
...
+    std::reverse( renderers.begin(), renderers.end() );
+  }
+  return renderers;
+}
+
+void QgsStackedDiagramRenderer::addRenderer( QgsDiagramRenderer *renderer )
+{
+  if ( renderer )
+  {
+    mDiagramRenderers.append( renderer );
```
Comment (nyalldawson) [inline src/core/qgsdiagramrenderer.cpp:1045]: @gacarrillor What is supposed to happen with ownership here? Right now mDiagramRenderers is never cleaned up, so these are all leaked. There's also no SIP_TRANSFER annotation on the function declaration, so you'll get a crash if you call this method from python.

There's also a (very bad) ownership issue with clone() here, because that calls a default copy constructor which just directly copies mDiagramRenderers to the clone, even though all the pointers in that container belong to the original instance. 

Can you please address this ASAP? We are closing in on release and this API is broken in its current state...

Comment (gacarrillor) [reply src/core/qgsdiagramrenderer.cpp:1045]: I'll have a look at it, thanks a lot for pointing this out!

Comment (gacarrillor) [reply src/core/qgsdiagramrenderer.cpp:1045]: Addressed in https://github.com/qgis/QGIS/pull/59123
